python/ray/air/execution/_internal/event_manager.py

```diff
@@ -163,40 +234,452 @@ def wait(
+            elif future in self._futures_to_actor_termination:
+                # Termination future resolved
+                tracked_actor = self._futures_to_actor_termination[future]
+                self._try_stop_tracked_actor(tracked_actor)
+            else:
+                # Resource future
+                self._handle_ready_resource_future()
+
+        # Lastly, if we have events left, try to start actors again (some resources
+        # might have been freed by our previous stop events).
+        if event_type in [EventType.ALL, EventType.ACTORS]:
+            if num_processed_events > 0 and num_events:
+                events_left = num_events - num_processed_events
+            else:
+                events_left = num_events
+
+            self._try_start_actors(max_actors=events_left)
+
+    def _handle_ready_actor_task_future(self, future: ray.ObjectRef):
+        """Handle an actor task future that became ready.
+
+        - Resolve future
+        - On actor error, trigger actor error callback AND error task error callback
+        - On task error, trigger actor task error callback
+        - On success, trigger actor task result callback
+        """
+        tracked_actor_task = self._futures_to_tracked_actor_tasks.pop(future)
+        tracked_actor = tracked_actor_task._tracked_actor
+        self._tracked_actors_to_futures[tracked_actor].remove(future)
+
+        try:
+            result = ray.get(future)
+        except RayActorError as exc:
+            # Here the actual actor process died.
+            # First, clean up any references to the actor and its futures
+            self._cleanup_failed_actor(tracked_actor)
+
+            # Actor failed - first trigger actor error callback
+            if tracked_actor._on_error:
+                tracked_actor._on_error(tracked_actor, exc)
+
+            # Then trigger actor task error callback
+            if tracked_actor_task._on_error:
+                tracked_actor_task._on_error(tracked_actor, exc)
+            return
+        except RayTaskError as exc:
+            # Here, only the task failed, but the actor process is still alive.
+            # Trigger actor task error callback
+            if tracked_actor_task._on_error:
+                tracked_actor_task._on_error(tracked_actor, exc)
+            return
+
+        # Trigger actor task result callback
+        if tracked_actor_task._on_result:
+            tracked_actor_task._on_result(tracked_actor, result)
+
+    def _handle_ready_resource_future(self):
+        """Handle a resource future that became ready.
+
+        - Update state of the resource manager
+        - Try to start one actor
+        """
+        # Force resource manager to update internal state
+        self._resource_manager.update_state()
+        # We handle resource futures one by one, so only try to start 1 actor at a time
+        self._try_start_actors(max_actors=1)
+
+    def _try_start_actors_timeout(
```

Comment:
Here we implement our own timeout for starting actors in the case where we don't have any futures to await. If we have futures to await, we do that instead.